core/src/main/java/com/linecorp/armeria/server/AbstractBindingBuilder.java

```diff
@@ -0,0 +1,316 @@
+import static com.linecorp.armeria.common.HttpMethod.DELETE;
+import static com.linecorp.armeria.common.HttpMethod.GET;
+import static com.linecorp.armeria.common.HttpMethod.HEAD;
+import static com.linecorp.armeria.common.HttpMethod.OPTIONS;
+import static com.linecorp.armeria.common.HttpMethod.PATCH;
+import static com.linecorp.armeria.common.HttpMethod.POST;
+import static com.linecorp.armeria.common.HttpMethod.PUT;
+import static com.linecorp.armeria.common.HttpMethod.TRACE;
+import static com.linecorp.armeria.server.HttpHeaderUtil.ensureUniqueMediaTypes;
+import static java.util.Objects.requireNonNull;
+
+import java.util.EnumSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.MediaType;
+
+/**
+ * An abstract builder class binding something to a {@link Route} fluently.
+ */
+abstract class AbstractBindingBuilder {
+
+    private Set<HttpMethod> methods = ImmutableSet.of();
+    private Set<MediaType> consumeTypes = ImmutableSet.of();
+    private Set<MediaType> produceTypes = ImmutableSet.of();
+    private final Map<RouteBuilder, Set<HttpMethod>> routeBuilders = new LinkedHashMap<>();
+    private final Set<RouteBuilder> pathBuilders = new LinkedHashSet<>();
+
+    /**
+     * Sets the path pattern that a {@link Service} will be bound to.
+     * Please refer to the <a href="https://line.github.io/armeria/server-basics.html#path-pattern">Path pattern</a>
+     * in order to learn how to specify a path pattern.
+     *
+     * @throws IllegalArgumentException if the specified path pattern is invalid
+     */
+    public AbstractBindingBuilder path(String pathPattern) {
+        pathBuilders.add(Route.builder().path(requireNonNull(pathPattern, "pathPattern")));
+        return this;
+    }
+
+    /**
+     * Sets the specified prefix which is a directory that a {@link Service} will be bound under.
+     * {@code pathUnder("/my/path")} is identical to {@code path("prefix:/my/path")}.
+     *
+     * @throws IllegalArgumentException if the specified path pattern is invalid
+     * @deprecated Use {@link #prefix(String)}.
+     */
+    @Deprecated
+    public AbstractBindingBuilder pathUnder(String prefix) {
+        return prefix(prefix);
+    }
+
+    /**
+     * Sets the specified prefix which is a directory that a {@link Service} will be bound under.
+     * {@code prefix("/my/path")} is identical to {@code path("prefix:/my/path")}.
+     *
+     * @throws IllegalArgumentException if the specified path pattern is invalid
+     */
+    public AbstractBindingBuilder prefix(String prefix) {
```

Comment:
I thought we were supposed to use `pathPrefix`.

Reply:
I like the name! Thanks for a nice idea.🙇‍♂️